include/aspect/particle/particle_handler.h

```diff
@@ -128,6 +139,50 @@ namespace aspect
         remove_particle(const particle_iterator &particle);
 
         /**
+         * Insert a particle into the collection of particles. Return an iterator
+         * to the new position of the particle.
+         */
+        particle_iterator
+        insert_particle(const Particle<dim,spacedim> &particle,
+                        const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell);
+
+        /**
+         * This function allows to register three additional functions that are
+         * called every time a particle is transferred to another process
+         * (i.e. during sorting into cells, during ghost particle transfer, or
+         * during serialization of all particles.
+         *
+         * @param size_callback A function that is called when serializing
+         * particle data. The function gets no arguments and is expected to
+         * return the size of the additional data that is serialized per
+         * particle. Note that this currently implies the data size has to be
+         * the same for every particle, but it does not have to be the same for
+         * every serialization process (e.g. a serialization during particle
+         * movement might include temporary data, while a serialization after
+         * movement was finished does not need to transfer this data).
+         * @param store_callback A function that is called once per particle
+         * when serializing particle data. Arguments to the function are a
+         * particle iterator that identifies the current particle and a void
+         * pointer that points to a data block of size size_callback() in which
+         * the function can store additional data. The function is expected to
+         * return a void pointer pointing to a position right after its data
+         * block.
+         * @param load_callback A function that is called once per particle
+         * when deserializing particle data. Arguments to the function are a
+         * particle iterator that identifies the current particle and a void
+         * pointer that points to a data block of size size_callback() in which
+         * additional data was stored by the store_callback function. The
+         * function is expected to return a void pointer pointing to a position
+         * right after its data block.
+         */
+        void
+        register_additional_store_load_functions(const std_cxx11::function<size_t ()> &size_callback,
+                                                 const std_cxx11::function<void *(const particle_iterator,
+                                                                                  void *)> &store_callback,
+                                                 const std_cxx11::function<const void *(const particle_iterator,
+                                                                                        const void *)> &load_callback);
```

Comment:
Here's some food for thought, not necessarily meant as a request to change things but for you to think about: In the `hp::DoFHandler` work, I've made these things more flexible by just letting the store functions return a `std::vector<char>` (could also have been a `std::string`) that the store function can size and fill as it pleases. I then let the place that calls all of these callbacks deal with the question of how to pack this data into buffers. That was easy in that case because it did not have to squeeze things into the p4est framework of attaching data to cells, but I found it kind of liberating to do it this way around because the store and load functions become so much simpler.  It required a little bit more logic in one central place, where I call these functions, but I thought that was worth it. Could something like this be done here as well?